pkg/apis/pmemcsi/v1alpha1/deployment_types.go

```diff
@@ -0,0 +1,427 @@
+	// DefaultDriverImage default PMEM-CSI driver docker image
+	DefaultDriverImage = defaultDriverImageName + ":" + defaultDriverImageTag
+
+	defaultProvisionerImageName = "quay.io/k8scsi/csi-provisioner"
+	defaultProvisionerImageTag  = "v1.2.1"
+	// DefaultProvisionerImage default external provisioner image to use
+	DefaultProvisionerImage = defaultProvisionerImageName + ":" + defaultProvisionerImageTag
+
+	defaultRegistrarImageName = "quay.io/k8scsi/csi-node-driver-registrar"
+	defaultRegistrarImageTag  = "v1.1.0"
+	// DefaultRegistrarImage default node driver registrar image to use
+	DefaultRegistrarImage = defaultRegistrarImageName + ":" + defaultRegistrarImageTag
+
+	// DefaultControllerResourceCPU default CPU resource limit used for controller pod
+	DefaultControllerResourceCPU = "100m" // MilliSeconds
+	// DefaultControllerResourceMemory default memory resource limit used for controller pod
+	DefaultControllerResourceMemory = "250Mi" // MB
+	// DefaultNodeResourceCPU default CPU resource limit used for node driver pod
+	DefaultNodeResourceCPU = "100m" // MilliSeconds
+	// DefaultNodeResourceMemory default memory resource limit used for node driver pod
+	DefaultNodeResourceMemory = "250Mi" // MB
+	// DefaultDeviceMode default device manger used for deployment
+	DefaultDeviceMode = DeviceModeLVM
+)
+
+var (
+	// DefaultNodeSelector default node label used for node selection
+	DefaultNodeSelector = map[string]string{"storage": "pmem"}
+)
+
+// DeploymentPhase represents the status phase of a driver deployment
+type DeploymentPhase string
+
+const (
+	// DeploymentPhaseNew indicates a new deployment
+	DeploymentPhaseNew DeploymentPhase = ""
+	// DeploymentPhaseInitializing indicates deployment initialization is in progress
+	DeploymentPhaseInitializing DeploymentPhase = "Initializing"
+	// DeploymentPhaseRunning indicates that the deployment was successful
+	DeploymentPhaseRunning DeploymentPhase = "Running"
+	// DeploymentPhaseFailed indicates that the deployment was failed
+	DeploymentPhaseFailed DeploymentPhase = "Failed"
+)
+
+// DeploymentChange type declaration for changes between two deployments
+type DeploymentChange int
+
+const (
+	DriverName DeploymentChange = iota + 1
+	DriverMode
+	DriverImage
+	PullPolicy
+	LogLevel
+	ProvisionerImage
+	NodeRegistrarImage
+	ControllerResources
+	NodeResources
+	NodeSelector
+)
+
+func (c DeploymentChange) String() string {
+	return map[DeploymentChange]string{
+		DriverName:          "driverName",
+		DriverMode:          "deviceMode",
+		DriverImage:         "image",
+		PullPolicy:          "imagePullPolicy",
+		LogLevel:            "logLevel",
+		ProvisionerImage:    "provisionerImage",
+		NodeRegistrarImage:  "nodeRegistrarImage",
+		ControllerResources: "controllerResources",
+		NodeResources:       "nodeResources",
+		NodeSelector:        "nodeSelector",
+	}[c]
+}
+
+// EnsureDefaults make sure that the deployment object has all defaults set properly
+func (d *Deployment) EnsureDefaults() {
```

Comment:
Is this hand-written or generated?

Reply:
This method is written by us _not_ generated by sdk.